adserver/tasks.py

```diff
@@ -490,6 +493,106 @@ def daily_update_uplift(day=None):
         )
 
 
+@app.task()
+def daily_update_advertisers(day=None):
+    """
+    Generate the daily index of AdvertiserImpressions.
+
+    :arg day: An optional datetime object representing a day
+    """
+    start_date, end_date = get_day(day)
+
+    log.info("Updating advertiser impressions for %s-%s", start_date, end_date)
+
+    # Important: uses the *already calculated* AdImpression index
+    # This should make this much faster than using the Offers table
+    queryset = AdImpression.objects.using(settings.REPLICA_SLUG).filter(
+        date__gte=start_date,
+        date__lt=end_date,  # Things at UTC midnight should count towards tomorrow
+    )
+
+    for values in (
+        queryset.values(
+            "advertisement__flight__campaign__advertiser__name",
+            "advertisement__flight__campaign__advertiser_id",
+        )
+        .annotate(
+            total_decisions=Sum("decisions"),
+            total_offers=Sum("offers"),
+            total_views=Sum("views"),
+            total_clicks=Sum("clicks"),
+            total_spend=Sum(
+                (F("clicks") * F("advertisement__flight__cpc"))
```

Comment:
I wish we had a nicer way to single source this stuff. It feels error prone to repeat it over & over, but it does tend to change slightly each time 🙃

Reply:
I think the only way to really do that is stored procedures but I think they'll differ a bit each time.